<a href="https://colab.research.google.com/github/kylampiima/Data-Analytics/blob/main/Genetic_Algorithm_Assignment_Kyla_Mpiima_2025.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Genetic Algorithm Assignment
30% of the overall grade for this module

Marks indciated in sections below are based on percentage of marks allocated for this module

In this assignment you must choose a problem, and attempt to use the Genetic Alogrithm that we developed in class to solve this problem.





## The Problem         **(~30%)**

*   Description of the problem





---

The task at hand is to create a synthetic audio waveform that sounds very similar to a specific target audio file. I want to generate a new sound that matches an existing sound as closely as possible.

Sounds are made up of waves that can be very complex. Each sound has many different characteristics, like pitch and volume, which makes it hard to recreate accurately.

The goal is to minimize the difference between our generated sound and the target sound. This difference is measured using a cost function, which tells us how close we are to achieving our goal. The smaller the difference, the better our generated sound matches the target.

---



*   Discussion of the suitablity of Genetic Algorithms


---

Exploration of Options: Just like in nature, GAs can explore many different combinations of sound waves to find the best match. This is important because the space of possible sounds is huge.

Adaptation: GAs can adapt and improve over generations. If one sound is closer to the target, it can be used to create new sounds that might be even better.

Avoiding Dead Ends: By keeping a diverse group of sounds, GAs can avoid getting stuck on a solution that isn’t the best, allowing them to find better matches.

In summary, the problem is about generating a new sound that closely resembles a given target sound, and using a Genetic Algorithm helps us explore and refine our solutions effectively.

---



*   Complexity of the problem  (Overall marks allocated based on ..)

# The problem and the cost function   **(~20%)**

---

The cost function used in this project is the sum of squared differences between the generated audio waveform and the target audio waveform. This approach was selected for its sensitivity to larger discrepancies, which is crucial in audio synthesis, as even minor deviations can significantly affect sound quality. By squaring the differences, the function emphasizes substantial errors, ensuring that the Genetic Algorithm (GA) focuses on minimizing these discrepancies for a more accurate match.

Additionally, the mathematical simplicity of this formulation allows for efficient computation, enabling rapid evaluation of candidate solutions during the GA's iterative process. This efficiency is vital given the large size of audio data, facilitating quicker convergence towards an optimal solution. The smooth and continuous nature of the squared difference function is also advantageous for optimization, providing a clear gradient for improvement and helping to avoid erratic behavior in the GA.

Furthermore, the cost function's robustness to noise allows it to average out variations in the audio data, leading to a more reliable measure of similarity. This is particularly important in achieving high-quality matches despite imperfections in the target audio. Overall, the choice of the squared difference cost function is a critical element in guiding the GA effectively towards generating high-fidelity audio waveforms that closely resemble the target sound.

---



In [ ]:
import numpy as np
from scipy.io.wavfile import read, write
import io
# Cost Function
def differenceBetweenWav(x):


  with open("output_sample_1000.wav", "rb") as wavfile:
        input_wav = wavfile.read()

  rate, data = read(io.BytesIO(input_wav))

  audio_data = data

  difference = 0
  difference = sum((x- audio_data)**2)
  return difference

**Testing the input wav file**

> This code snippet reads an audio file (output_sample_1000.wav) in binary format and loads it into the variable input_wav. It then uses the read function  to extract the sample rate (rate) and audio data (data) from the WAV file.

> The variable tester_data is assigned the audio data (data). The commented-out lines show operations, finding the minimum and maximum values of the data or checking the sample rate.

> The duration of the audio is calculated by dividing the total number of samples (len(data)) by the sample rate (rate), giving the length of the audio in seconds. Finally, the length of tester_data is determined using len(tester_data), which gives the total number of samples in the audio.



In [ ]:
with open("output_sample_1000.wav", "rb") as wavfile:
        input_wav = wavfile.read()

rate, data = read(io.BytesIO(input_wav))

tester_data = data

#tester_data.min()
#tester_data.max()
#rate
duration = len(data) / rate
duration
#len(tester_data)



1.6

The high number of genes makes this a complex system, potentially modeling a high-dimensional space where each gene represents a unique parameter or feature. The acceptable_cost threshold of 0.000000001  the precision required for the solution, showing that the problem demands extremely fine accuracy.

In [ ]:
class problem:
  def __init__(self):
    self.number_of_genes = 25600
    self.max_value = 4572
    self.min_value = -4686
    self.cost_function = differenceBetweenWav
    self.acceptable_cost = 0.000000001
    self.rate = 16000
    self.duration = 1.6

# The Individual **(~30%)**


*   Chromosone
*   Crossover
*   Mutation



## Discussion and justification on the approaches taken for the above

the key aspects of the individual class is its crossover method, which combines genetic material from two parent individuals to produce offspring. In this implementation, a fixed crossover point is used to splice the chromosomes of the two parents. Specifically, the method joins the beginning segment of one parent's chromosome with the ending segment of the other parent. This approach is particularly important for audio waveforms, as the sequential nature of audio data means that maintaining the order of samples is crucial for preserving sound quality.

Traditional crossover methods, which might randomly swap segments of chromosomes, are not suitable for audio data because they can disrupt the  coherence necessary for the waveform to sound natural. By using a fixed crossover point, the method ensures that the resulting offspring inherit a continuous segment of audio data from each parent, thereby maintaining the integrity of the waveform. This is essential for generating audio that is not only similar to the target but also coherent and listenable


In [ ]:
class individual:
    def __init__(self, prob):
      self.chromosome = np.random.uniform(prob.min_value, prob.max_value,prob.number_of_genes)
      self.cost = prob.cost_function(self.chromosome)
    def mutate(self, rate_of_gene_mutation, range_of_gene_mutation):
        for index in range(len(self.chromosome)):
            if np.random.uniform() < rate_of_gene_mutation:
                self.chromosome[index] += np.random.randint(-100, 101) * range_of_gene_mutation

    def crossover(self, parent2, explore_crossover):
      crossoverPoint = 1250; #random midpoint
      child1 = deepcopy(self)
      child2 = deepcopy(parent2)
      child1.chromosome = np.append(self.chromosome[:crossoverPoint] , parent2.chromosome[crossoverPoint:])
      child2.chromosome = np.append( parent2.chromosome[:crossoverPoint] , self.chromosome[crossoverPoint:])

      return child1, child2

## Running the algorithm  **(~10%)**

*   Parameter choices
*   Modifications (if any) to run_genetic
*   Rationale for the above



---

A population size of 500 ensures sufficient genetic diversity, while an explore crossover range of 0.1 allows for moderate innovation without compromising the integrity of the audio signal. The gene mutation range of 0.5 enables significant adjustments to individual genes, facilitating the discovery of better solutions, and a mutation rate of 1 ensures continuous variability across the population. With a birth rate of 2, successful traits are effectively passed on, and the maximum of 500 generations strikes a balance between exploration and computational efficiency. Together, these parameters create a robust framework that promotes effective evolution of high-quality audio waveforms, making this a strong starting point for the project.

---



In [ ]:

class parameters:
  def __init__(self):
    self.population  = 1000
    self.explore_crossover_range = 0.2
    self.gene_mutate_range = 1
    self.gene_mutate_rate = 0.2
    self.birth_rate_per_generation = 1
    self.max_number_of_generations = 1000


In [ ]:
def choose_parents(population):
  index1 = np.random.randint(0,len(population))
  index2 = np.random.randint(0,len(population))
  if index1 == index2:
    return choose_parents(population)
  return index1, index2

In [ ]:
from copy import deepcopy

def run_genetic(prob,params):
  number_in_population = params.population
  explore_crossover = params.explore_crossover_range
  gene_mutate_rate = params.gene_mutate_rate
  max_number_of_generations = params.max_number_of_generations
  gene_mutate_range = params.gene_mutate_range
  cost_function = prob.cost_function
  number_of_children_per_generation = number_in_population * params.birth_rate_per_generation
  acceptable_cost = prob.acceptable_cost

  population = []
  best_solution = individual(prob)
  best_solution.cost = np.inf
  for i in range(number_in_population):
    new_individual = individual(prob)
    if new_individual.cost:
      best_solution = deepcopy(new_individual)
    population.append(individual(prob))

  for i in range(max_number_of_generations):
    children = []

    while len(children) < number_of_children_per_generation:
      parent1_index, parent2_index = choose_parents(population)

      parent1 = population[parent1_index]
      parent2 = population[parent2_index]

      child1, child2 = parent1.crossover(parent2,explore_crossover)
      child1.mutate(gene_mutate_rate, gene_mutate_range)
      child1.cost = cost_function(child1.chromosome)
      child2.mutate(gene_mutate_rate, gene_mutate_range)
      child2.cost = cost_function(child2.chromosome)

      children.append(child1)
      children.append(child2)

    population += children

    population = sorted(population, key=lambda x: x.cost)

    population = population[:number_in_population]

    if population[0].cost < best_solution.cost:
      best_solution = deepcopy(population[0])
      print(best_solution.cost)
    if best_solution.cost < acceptable_cost:
      break

  return population, best_solution

The saveWav function is crucial for exporting the best audio solution generated by a genetic algorithm into a WAV file format, enabling playback and analysis. It calculates the total number of samples based on the sample rate and duration, initializes an audio data array filled with the best solution's cost, and clips the data to ensure it remains within the valid range for 16-bit PCM audio. The function then converts the data to the appropriate type and writes it to a WAV file using the write function. Finally, it provides user feedback to confirm successful file creation. Overall, this function ensures that the generated audio is preserved in a standard format while maintaining quality, making it essential for evaluating the algorithm's performance.

In [ ]:
def saveWav(best_solution,sample_rate, duration, output_filename="modified.wav"):
    num_samples = int(sample_rate * duration)
    audio_data = np.full(num_samples, best_solution.cost, dtype=np.float64)


    audio_data = np.clip(audio_data, -32768, 32767)
    audio_data = audio_data.astype(np.int16)

    write(output_filename, sample_rate, audio_data)

    print(f"WAV file '{output_filename}' created ")

## **Testing and Evaluation**

**To understand how well the system performs and how the chosen parameters influence results, a series of tests were carried out. These tests aimed to evaluate the effectiveness of the evolutionary approach in generating high-quality audio outputs. Different settings were explored, and the behavior of the system was closely observed to assess performance, stability, and the quality of evolved solutions. The goal was to identify what works best and where improvements could be made.**

In [ ]:
prob1 = problem()
par1 = parameters()

In [ ]:
best = individual(prob1)
print("new_individual.cost:", best )
print(best)

new_individual.cost: <__main__.individual object at 0x7d69d10c5c90>


In [ ]:
pop, best = run_genetic(prob1,par1)

14171152000.0
10012219000.0
8249318400.0
7136657400.0
6446893000.0
5963190300.0
5712762000.0
5541174000.0
5401733000.0
5339003000.0
5295373300.0
5264958500.0
5243541500.0
5223386000.0
5220849700.0
5213248500.0
5212463000.0
5211040000.0
5209617400.0
5208754000.0
5206564000.0
5206025000.0


KeyboardInterrupt: 

The results show a significant and consistent decrease in the cost, starting from 14171152000.0 and steadily dropping to 5206025000.0. This indicates that the algorithm is successfully minimizing the difference between the generated sound and the target sound over time.

The reduction in cost is quite pronounced initially, suggesting that the algorithm was able to make substantial improvements in the early stages. However, as the values get lower, the rate of improvement slows down, which is common as the algorithm reaches a more refined solution space.

The relatively small reductions in the later stages of the process (from 5211040000.0 to 5206025000.0) indicate that the algorithm is approaching a point where further significant improvements are harder to achieve. This could imply that the algorithm has effectively converged toward a near-optimal solution, but is struggling to improve further.

In [ ]:
best.cost

np.float32(13972073000.0)

In [ ]:
saveWav(best,prob1.rate,prob1.duration)

WAV file 'modified.wav' created 


In contrast to the first configuration, this set of parameters offers a more focused and efficient approach to evolution. While the first setup increases population size and generations, potentially adding overhead, this maintains a manageable population of 500 and caps generations at 500—allowing for quicker iteration and evaluation. The higher mutation rate of 1 in this setup ensures continuous variation, helping avoid early stagnation, while the second’s lower rate (0.2) may slow down exploration. Similarly, the more moderate mutation range (0.5) in this configuration allows for meaningful changes without destabilizing good solutions, unlike the more aggressive range of 1 in the second. With a higher birth rate of 2, this setup also promotes successful traits more actively, encouraging faster convergence. Overall, this makes this configuration more effective in generating strong results within a shorter and more computationally efficient window.

In [ ]:
class parameters:
  def __init__(self):
    self.population  = 500
    self.explore_crossover_range = 0.1
    self.gene_mutate_range = 0.5
    self.gene_mutate_rate = 1
    self.birth_rate_per_generation = 2
    self.max_number_of_generations = 500

In [ ]:
p1 = individual(prob2)
p2 = individual(prob2)
(c1,c2) = p1.crossover(p2,par2.explore_crossover_range)

In [ ]:
c1.cost

np.float64(244612872732.94086)

In [ ]:
prob2 = problem()
par2 = parameters()

In [ ]:
pop, best = run_genetic(prob2,par2)

187213470611.58655
186347656195.89618
185620298385.10593
185531674808.00214
185514561277.31036
185439829790.54843
185378992806.57767
185351868590.25537
185343822274.61154
185313985266.48123
185281100493.39008
185256584493.4616
185240059925.39215
185216099895.90045
185214313550.51862
185158233269.1148
185133087731.28085
185092941365.038
185083601393.42374
185066253173.58057
185031361050.05847
184993733832.12656
184936623888.07413
184929546261.08334
184915052198.05173
184907299809.95737
184845037220.71625
184841708883.9871
184829484005.08978
184801195521.63626
184768111881.59204
184728951338.53354
184693095927.3192
184676916903.00046
184600379624.6208
184536801291.01428
184518783116.6994
184496694757.88248
184469729919.6798
184428565345.76584
184387476700.44183
184346249138.20358
184328892440.53107
184270971732.31503
184267220796.1545
184252694758.52148
184221592245.14468
184167489279.16016
184160289839.57913
184117017741.84274
184100172665.44058
184083974463.41525
184067429071.62396
184

The results show a gradual decrease in the cost, starting from 187213470611.58655 and slowly dropping to 180790746980.85736. This indicates that the algorithm is making progress towards reducing the difference between the generated sound and the target sound, though the improvements are slow.

However, the reduction in cost is not consistent throughout the process. At times, the values barely change or even increase slightly, which suggests that the algorithm may be getting stuck in certain areas and not finding better solutions.

While the overall trend shows improvement, the rate of progress slows down as the algorithm continues. This could indicate that the algorithm has explored most of the better solutions within its current configuration, and further significant improvements are more difficult to achieve.

In [ ]:
class parameters:
    def __init__(self):
        self.population = 300
        self.explore_crossover_range = 0.3
        self.gene_mutate_range = 0.1
        self.gene_mutate_rate = 0.2
        self.birth_rate_per_generation = 1
        self.max_number_of_generations = 500


Compared to the second parameter setup, this configuration takes a more conservative and exploratory approach. The **population size is smaller (300 vs 500)**, which reduces computational load but may limit genetic diversity and slow overall progress. However, the **higher crossover range (0.3 vs 0.1)** encourages more exploration and mixing of traits, possibly uncovering novel solutions, though it risks losing stability.

On the mutation side, this version opts for **smaller, more cautious changes**—a **lower mutation range (0.1 vs 0.5)** and **lower mutation rate (0.2 vs 1)**—which may help fine-tune solutions but slows down variation and adaptability, especially early on. Additionally, with a **birth rate of just 1**, the spread of successful traits happens more gradually compared to the faster reinforcement seen with a birth rate of 2 in the second setup.

Overall, while the second configuration leans toward faster evolution with greater variability and stronger trait propagation, this third configuration trades speed for **stability and control**, potentially making it more suited for fine-tuning later stages of evolution or avoiding drastic changes that destabilize performance.

In [ ]:
prob3 = problem()
par3 = parameters()

In [ ]:
pop, best = run_genetic(prob3,par3)

186877945939.63113
186647194516.13837
186633843136.79666
186625840088.21817
186623877333.08624
186620248330.90286
186617113965.0836
186611779389.57098
186607601865.80078
186604032005.15363
186601261761.09088
186597647803.12396
186594127446.75397
186589635202.06842
186584927450.20474
186581223678.88733
186576962856.5475
186570758721.72025
186568253575.06778
186565278662.60184
186559851061.68826
186556148541.31357
186551205763.3396
186546782672.8828
186544280553.77484
186540137987.0598
186536455651.08344
186531226409.61398
186525974908.19492
186524705756.80722
186518188682.01895
186514932249.8969
186509029600.27634
186507470569.1818
186502606471.59024
186498884115.2302
186491578911.5626
186489957007.5172
186485789450.17245
186481340365.875
186474992188.09464
186474069943.16434
186468529284.45074
186464171523.88647
186459263093.92648
186457148819.74292
186452669005.17032
186448892818.21542
186443384506.67572
186439464572.93317
186436451239.83975
186433475816.50714
186428508617.97037
18642

**Failed to run to completion**

In [ ]:
par4 = parameters()
prob4 = problem()

In [ ]:
pop, best = run_genetic(prob4,par4)

186827193273.07352
186753470943.76733
186644676317.5989
186638941482.54404
186633809823.5802
186629944080.4973
186627666988.372
186624963705.41995
186618493335.77948
186616555257.7708
186612521916.4849
186606751079.56992
186605139713.80716
186600289475.61215
186595909475.9285
186590370178.6719
186586199602.03604
186582830353.81778
186578878897.64673
186575500276.61426
186570514986.09332
186565700406.735
186564333637.2812
186556901769.449
186554979330.34354
186548857696.02896
186546340000.00537
186542585854.74332
186540259468.77432
186533556026.13528
186526986025.0474
186525679023.90305
186521209518.84363
186516519158.4524
186513233368.02347
186508550963.13507
186503622567.13907
186500674576.20905
186495830339.6871
186491123456.38577
186488721159.46817
186483597867.02914
186478837216.21536
186473961754.11725
186468521251.87833
186465834309.9989
186463000596.0776
186458836599.57843
186454940092.05963
186450831349.73065
186446198839.48074
186442994754.79556
186437631802.3593
186432848519.

KeyboardInterrupt: 

**Failed to run to completion**

In [ ]:
par5 = parameters()
prob5 = problem()

In [ ]:
pop, best = run_genetic(prob5,par5)

187490191529.603
187237053686.4275
187165513998.89978
187056939886.0457
187056674159.99063
186982254018.13943
186980516788.5907
186976743855.1323
186973261772.67453
186972556212.7965
186969696259.775
186968811952.8422
186965415876.531
186961798365.3238
186961409160.62347
186961139600.9837
186957236703.62128
186956380632.84775
186953321605.73358
186947901171.9103
186945167115.82162
186943684902.44458
186940870305.69427
186937287553.82675
186934260958.81912
186932761932.6883
186930939976.61798
186926069112.26355
186924380142.47153
186922471085.48376
186920841235.97903
186918288846.81393
186912977727.05
186911320753.2173
186908230290.01227
186907924404.91974
186903781211.7435
186901381762.59763
186896384808.16037
186893272418.28568
186893150245.34775
186889736386.7209
186884673773.41028
186884440175.10184
186879646663.2766
186878748845.06485
186875562661.97852
186875287201.36597
186870997032.5437
186868529428.55313
186866021159.3968
186864973976.47293
186860560312.51132
186858079233.69366

The results show a steady decrease in values, starting at 187490191529.603 and gradually dropping to 185967807084.74594. This suggests that the process or system being measured is making incremental progress toward a goal, with a consistent reduction in the values recorded.

Initially, the changes from one value to the next are relatively large, indicating that the system was adjusting or improving significantly in the earlier stages. However, as the values decrease, the changes become smaller, which typically means that the system is approaching a point of diminishing returns—where further improvements are harder to achieve.

The slowdowns in the rate of decrease in the later stages suggest that the system is nearing an optimal point, where any further progress will be incremental rather than significant. This could indicate convergence toward a steady state, meaning the system is stabilizing and cannot improve much more.



In [ ]:
saveWav(best,prob5.rate,prob5.duration)

WAV file 'modified.wav' created 


## Results and conclusions    **(~10%)**

The first and second sets of tests, which used larger population sizes and different crossover and mutation settings, did not complete due to the excessive time required to run. Specifically, these tests took more than 12 hours to process without reaching completion, which indicates that the larger population sizes and other parameter choices were computationally expensive. While these setups may have had the potential to provide more diverse solutions or more substantial optimization, the computational cost rendered them impractical for this task within the given time constraints.

The fact that the third set of tests eventually provided results, albeit with slow improvement, suggests that the genetic algorithm could be further optimized by fine-tuning the population size, crossover range, and mutation rate. To improve performance, it may be beneficial to either reduce the population size further or increase the mutation rate to encourage more variability and exploration. Additionally, adjusting the crossover range to optimize the balance between exploration and exploitation could lead to faster convergence and better results within a reasonable timeframe.

Given the computational difficulties experienced with the first and second sets of tests, a focus on optimizing parameter efficiency will be key to ensuring that the genetic algorithm performs effectively without excessively long runtimes.

---

In conclusion, the analysis of results from using a genetic algorithm to minimize the difference between generated sound and the target sound shows some improvement over time. The algorithm is able to reduce the cost, which measures how closely the generated sound matches the target. By adjusting the parameters the algorithm has become better at exploring different solutions and fine-tuning the generated sound.

However, there are challenges with this approach. The algorithm is taking a long time to complete, and the improvements in the cost are not happening quickly. This suggests that while the algorithm is making progress, it is doing so at a slow pace, possibly because it is not finding the most effective solutions efficiently. This slow progress could be due to conservative settings for mutation and crossover, or because the solution space is too large and difficult to navigate with the current parameters.

Additionally, the large amount of data required to generate and match the sound adds to the time needed for processing. Genetic algorithms, while powerful in exploring solutions, can be slow and require a lot of computing power, especially when dealing with complex tasks like sound matching, where even small changes in input can lead to big differences in output.

To improve performance and reduce the time required, it might help to adjust the parameters further. Increasing the population size could lead to more diverse solutions and speed up the search process, although this would also increase computational demands. Other methods, such as parallel processing or more efficient selection and mutation strategies, could help the algorithm find solutions faster. Additionally, using different optimization techniques, like gradient-based methods or combining genetic algorithms with other approaches, might speed up the process without losing quality in the results.

In summary, although the genetic algorithm shows potential for optimizing sound generation, it needs further adjustments to address the slow progress and high computational demands. By fine-tuning the settings and exploring other methods, the optimization process could become faster and more effective.



---

